File: 73.set-matrix-zeroes.cpp
/*
 * @lc app=leetcode id=73 lang=cpp
 *
 * [73] Set Matrix Zeroes
 */
#include <vector>

using std::vector;

// @lc code=start
class Solution {
public:
    // T: O(m * n)
    // S: O(1)
    void setZeroes(vector<vector<int>>& matrix) {
        const int m = matrix.size(); 
        const int n = matrix[0].size();
        bool col_zero_flag = false; 
        bool row_zero_flag = false;
        // Check if 1st row contains zero
        for (int j = 0; j < n; ++j) {
            if (matrix[0][j] == 0) {
                row_zero_flag = true;
                break;
            }
        }
        // Check if 1st col contains zero
        for (int i = 0; i < m; ++i) {
            if (matrix[i][0] == 0) {
                col_zero_flag = true;
                break;
            }
        }

        // Use 1st row and 1st col to store whether the corresponding 
        // cols/rows should be zeroed (with the bool flags, information 
        // not lost)
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (matrix[i][j] == 0) {
                    // if (i == 0) {
                    //     row_zero_flag = true;
                    // }
                    // if (j == 0) {
                    //     col_zero_flag = true;
                    // }
                    matrix[i][0] = matrix[0][j] = 0;
                }
            }
        }

        // Process cells except 1st row and 1st col
        for (int i = 1; i < m; ++i) {
            for (int j = 1; j < n; ++j) {
                if (matrix[i][0] == 0 || matrix[0][j] == 0) {
                    matrix[i][j] = 0;
                }
            }
        }

        // Process 1st row and 1st col
        if (row_zero_flag) {
            for (int j = 0; j < n; ++j) {
                matrix[0][j] = 0;
            }
        }
        if (col_zero_flag) {
            for (int i = 0; i < m; ++i) {
                matrix[i][0] = 0;
            }
        }
    }
};
// @lc code=end

File: 73.set-matrix-zeroes.go
/*
 * @lc app=leetcode id=73 lang=golang
 *
 * [73] Set Matrix Zeroes
 */

// @lc code=start
// T: O(m * n)
// S: O(1)
func setZeroes(matrix [][]int) {
	m := len(matrix)
	n := len(matrix[0])

	row0Flag := false
	col0Flag := false
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if matrix[i][j] == 0 {
				// Use the first row and the first col to mark existence(s) of zero.
				matrix[i][0] = 0
				matrix[0][j] = 0

				// Use a row flag and a col flag to check if the first row and/or the first column
				// should also be zeroed after processing all other rows and columns.
				if i == 0 {
					row0Flag = true
				}

				if j == 0 {
					col0Flag = true
				}
			}
		}
	}

	// Process entries except the first row and the first column based on
	// info stored in the first row and the first column
	for i := 1; i < m; i++ {
		for j := 1; j < n; j++ {
			if matrix[i][0] == 0 || matrix[0][j] == 0 {
				matrix[i][j] = 0
			}
		}
	}

	// Now zero the first row and/or the first column if needed
	if row0Flag {
		for j := 0; j < n; j++ {
			matrix[0][j] = 0
		}
	}

	if col0Flag {
		for i := 0; i < m; i++ {
			matrix[i][0] = 0
		}
	}
}

// @lc code=end

